### PR TITLE
feat: fail-closed driver execution — audit, deadlines, pooled HTTP client

### LINKED ISSUE #152: Guarantee an audit trace and budget release on any driver exception
## Summary

Ensure every invocation that reaches driver execution records an `ActionTrace` (success
or failure) and releases any reserved budget, even when a driver raises an exception
type other than `DriverError`.

## Why this matters

I-02 requires that every execution be auditable. `execute_with_fallback` only catches
`DriverError`; a `ValueError`, `TimeoutError`, `httpx` decode error, or any unexpected
exception propagates out of `perform_invoke` with no failure trace recorded and the
`BudgetManager` reservation never released. That is both an audit gap and a slow budget
leak under real-world driver failures.

## Current evidence

- `kernel/_invoke.py`: `execute_with_fallback` has `except DriverError as exc:` only; other exceptions are not caught.
- The budget release path (`kernel.budget.release(...)`) runs only inside the `raw_result is None` branch and the `finally` around the firewall transform — an exception during `driver.execute` bypasses both before a trace is written.
- `record_failure_trace` exists but is only invoked when drivers return `None`, not when they raise unexpectedly.

## External context

Not required for this issue.

## Proposed implementation

1. In `execute_with_fallback`, decide policy for non-`DriverError` exceptions: either
   wrap-and-continue to the next driver (treat as a failed attempt) or re-raise after
   recording — but in all cases ensure a failure `ActionTrace` is written and the
   reservation released.
2. Wrap the driver-execution-through-firewall span in `perform_invoke` so any escape
   records a failure trace and releases budget exactly once.
3. Preserve the existing `DriverError` aggregation/last-error message behavior.

## AI-agent execution notes

- Inspect first: `kernel/_invoke.py` (`execute_with_fallback`, `perform_invoke`, `record_failure_trace`), `firewall/budget_manager.py`, `tests/test_kernel.py`, `tests/test_firewall.py`.
- Edge cases: multiple drivers in a route plan where one raises unexpectedly and a later one succeeds — the success path must still record exactly one trace.
- Do not swallow the original exception's information; preserve it in the trace `error`.

## Acceptance criteria

- A driver raising a non-`DriverError` exception produces a failure `ActionTrace` retrievable via `explain()`.
- Reserved budget is released exactly once on every failure path.
- Existing `DriverError` fallback behavior is unchanged.

## Test plan

Add tests with a driver that raises `RuntimeError` and one that raises mid-stream; assert
trace recorded, budget restored, and (where applicable) fallback to the next driver. Run `make ci`.

## Documentation plan

Note the guarantee in `docs/architecture.md` invoke-pipeline description; CHANGELOG `Fixed`.

## Migration and compatibility notes

Unexpected exceptions may now surface as `DriverError` (or be recorded then re-raised).
Document the chosen contract. Minor behavior change.

## Risks and tradeoffs

Broad `except Exception` must be deliberate and narrow in scope to avoid masking
programming errors; re-raise after recording where appropriate.

## Suggested labels

reliability, security, testing

### LINKED ISSUE #191: Enforce a kernel-level invocation deadline across all drivers
## Summary

Add an overall per-invocation deadline enforced by the kernel (e.g.,
`asyncio.timeout` around driver execution), independent of driver-internal timeouts,
with the timeout recorded as a failure trace.

## Why this matters

Only `HTTPDriver` has its own timeout (`default_timeout=30.0`); `MCPDriver`, memory,
and third-party drivers can hang indefinitely, freezing the agent loop with a
budget reservation held. A kernel that promises controlled execution should bound
execution time as part of the contract, not delegate it to driver authors.

## Current evidence

- `drivers/http.py:39`: per-endpoint/driver timeout exists for HTTP only.
- `kernel/_invoke.py` `execute_with_fallback` awaits `driver.execute(...)` with no timeout wrapper; nothing in `kernel/__init__.py` bounds invocation duration.
- `drivers/base.py` driver Protocol has no timeout obligation.

## External context

Tool-execution gateways commonly enforce both per-call timeouts and global deadlines;
asyncio.timeout (3.11) / wait_for (3.10) are the standard primitives.

## Proposed implementation

1. Add `invoke_timeout_s: float | None` (kernel constructor or per-capability
   constraint), default None to preserve behavior, with a documented recommended value.
2. Wrap each driver attempt in `execute_with_fallback`; on timeout, synthesize a
   `DriverError("timeout after ...")` so existing fallback and failure-trace paths
   (and ISSUE 4's guarantees) apply.
3. Decide semantics: timeout per driver attempt vs whole route plan; per attempt is
   simpler and composes with fallback.
4. Streaming: apply an inactivity timeout between chunks in `kernel/_stream.py`.

## AI-agent execution notes

- Inspect first: `kernel/_invoke.py`, `kernel/_stream.py`, `drivers/base.py`, `tests/test_kernel.py`, `tests/test_drivers.py`.
- Python 3.10 support: use `asyncio.wait_for` (3.10-compatible), not `asyncio.timeout` (3.11+) — CI runs 3.10–3.12.
- Edge cases: cancellation must release budget reservations exactly once; timeout during fallback chain proceeds to next driver.
- Do not remove HTTPDriver's own finer-grained timeout.

## Acceptance criteria

- A driver sleeping past the deadline yields a failure trace with a timeout reason and releases budget.
- Fallback drivers are attempted after a timeout, matching `DriverError` semantics.
- Default (no timeout) behavior is byte-for-byte unchanged.

## Test plan

Async tests with a deliberately slow fake driver; timeout + fallback + budget-release
assertions; streaming inactivity case in `tests/test_firewall_stream.py`. Run `make ci`.

## Documentation plan

Document the deadline knob in `docs/architecture.md` and `docs/tutorial.md`;
CHANGELOG `Added`.

## Migration and compatibility notes

Opt-in; not expected to require migration.

## Risks and tradeoffs

Cancellation of an in-flight side-effecting call leaves remote state unknown —
document that timeout means "result discarded," not "action undone." Per-attempt
timeouts can extend total latency across long fallback chains.

## Suggested labels

reliability, product, testing

### LINKED ISSUE #194: Reuse a pooled HTTP client and add response-size limits in HTTPDriver
## Summary

Make `HTTPDriver` hold a long-lived `httpx.AsyncClient` (connection pooling,
configurable limits) instead of constructing one per request, and add an optional
maximum response size so unbounded bodies cannot exhaust memory before the firewall
ever sees them.

## Why this matters

A new `AsyncClient` per call means a fresh connection pool, TLS handshake, and no
keep-alive on every invocation — substantial avoidable latency for the most common
real driver. Reading unbounded response bodies also means the memory budget the
firewall enforces *after* parsing can be blown *before* it, by a single large
upstream response.

## Current evidence

- `drivers/http.py:96`: `async with httpx.AsyncClient(headers=headers, timeout=effective_timeout) as client:` inside `execute` — per-request client construction.
- No `httpx.Limits`, no `max_bytes`-style guard; `response.json()` parses the entire body regardless of size.
- `firewall/budget_manager.py` budgets apply only after `RawResult` is produced.

## External context

httpx documentation recommends a shared `AsyncClient` for connection pooling; the
per-request pattern is explicitly discouraged for performance.

## Proposed implementation

1. Create the client lazily in `__init__`/first use with headers and
   `httpx.Limits(...)`; add an async `aclose()` and document lifecycle (kernel or
   caller closes drivers on shutdown).
2. Per-endpoint header overrides can be passed per-request rather than per-client.
3. Add `max_response_bytes: int | None`; stream the body and abort with `DriverError`
   when exceeded.
4. Benchmark before/after with the local demo (`examples/http_driver_demo.py`).

## AI-agent execution notes

- Inspect first: `drivers/http.py`, `examples/http_driver_demo.py`, `tests/test_drivers.py`.
- Edge cases: event-loop lifecycle (client must be created on the loop that uses it); driver reuse across kernels; redirects and chunked encoding under the size cap.
- Preserve existing `DriverError` message formats — error messages are contract (AGENTS.md).
- Keep `timeout` semantics identical.

## Acceptance criteria

- One `AsyncClient` is reused across sequential invocations (asserted via transport mock).
- Responses above `max_response_bytes` raise `DriverError` without buffering the full body.
- Existing driver tests pass; `examples/http_driver_demo.py` still runs in `make example`.

## Test plan

Use `httpx.MockTransport` to count client/connection reuse; size-cap tests with
oversized fake bodies. Run `make ci`.

## Documentation plan

Update driver docs in `docs/integrations.md`/`docs/tutorial.md`; CHANGELOG
`Changed`/`Added`.

## Migration and compatibility notes

Behavioral note: connections are now kept alive; long-idle processes should call
`aclose()`. Default size cap of None preserves current behavior.

## Risks and tradeoffs

Client lifecycle management adds a small API surface (close/aclose). Pooling across
many distinct hosts has modest memory cost; `httpx.Limits` bounds it.

## Suggested labels

performance, reliability, integrations

### LINKED ISSUE #197: Treat malformed and non-JSON HTTP responses as DriverError
## Summary

Catch JSON decode failures in `HTTPDriver` (and optionally support non-JSON content
types explicitly) so a 200 response with a non-JSON body surfaces as a structured
`DriverError` instead of an uncaught `json.JSONDecodeError`.

## Why this matters

`HTTPDriver.execute` calls `response.json()` guarded only by `httpx.HTTPStatusError`
and `httpx.RequestError`. A successful response carrying HTML (e.g., a captive
portal, an error page, an API gateway timeout body) raises `JSONDecodeError`, which
is not a `DriverError` — so it skips fallback routing and, until ISSUE 4 lands,
escapes without a failure trace. This converts a mundane upstream quirk into an
audit gap and an agent-loop crash.

## Current evidence

- `drivers/http.py:109-118`: `data: Any = response.json()` inside the `try`, but the `except` clauses cover only `httpx.HTTPStatusError` and `httpx.RequestError`.
- `kernel/_invoke.py:133`: fallback catches `DriverError` only, so decode errors bypass fallback.
- AGENTS.md error contract: library code must raise typed kernel errors, not leak stdlib exceptions.

## External context

Not required for this issue.

## Proposed implementation

1. Wrap body parsing: catch `json.JSONDecodeError`/`ValueError` and raise
   `DriverError(f"HTTPDriver '...': non-JSON response from {url} (content-type: ...)")`.
2. Optionally add per-endpoint `response_format: "json" | "text"` so text APIs are
   supported deliberately (`RawResult.data` as string) rather than failing.
3. Include content-type and a bounded, redaction-safe body snippet in the error
   (coordinate with ISSUE 6 on scrubbing).

## AI-agent execution notes

- Inspect first: `drivers/http.py`, `tests/test_drivers.py` (existing MockTransport patterns).
- Preserve existing success-path behavior and error message style exactly.
- Edge cases: empty body with 204; JSON with BOM; content-type lies (JSON body with text/html header should still parse if `response_format="json"` succeeds).

## Acceptance criteria

- A 200 response with an HTML body raises `DriverError`, participates in fallback, and records a failure trace.
- `response_format="text"` (if implemented) returns the body as `RawResult.data`.
- No change to valid-JSON behavior.

## Test plan

MockTransport cases: HTML body, empty body, valid JSON, text mode. Run `make ci`.

## Documentation plan

Document endpoint response formats in `docs/integrations.md`; CHANGELOG `Fixed`.

## Migration and compatibility notes

Previously-crashing inputs now produce `DriverError`; not expected to require
migration.

## Risks and tradeoffs

Minimal. The optional text mode slightly widens the driver API; keep it small.

## Suggested labels

reliability, integrations, testing

### PR DESCRIPTION
**Pull request by dgenio**
Closes #152, closes #191, closes #194, closes #197.

These four issues were selected as a coherent group: they share one code area (`kernel/_invoke.py` + `drivers/http.py`) and one fail-closed implementation path — any fault funnels through `DriverError → fallback → failure trace → budget release`, so the kernel's "controlled, audited execution" promise (invariant **I-02**) holds on *every* exit, not just the happy path.

## What changed

**#152 — Audit + budget release on any driver fault**
- `execute_with_fallback` now treats *any* exception a driver raises (not only `DriverError`) as a failed attempt, captured as `last_error` instead of escaping un-audited with the budget reservation leaked.
- The post-driver pipeline in `perform_invoke` (handle creation, firewall transform, token counting) is wrapped in a fault-capturing `try/except`: a fault there releases the reservation **exactly once** and records a failure `ActionTrace` before re-raising.
- Extracted `execute_with_fallback` into a new `kernel/_driver_exec.py` so `_invoke.py` stays within the documented 300-line module budget; its size-ratchet ceiling is lowered `400 → 382` (shrink-only respected).

**#191 — Per-invocation deadline**
- Optional `invoke_timeout_s` token constraint bounds each driver attempt (single-shot and streaming, including a stream inactivity timeout) via `asyncio.wait_for` (3.10-compatible). A timeout becomes a synthetic `DriverError`, so the existing fallback + failure-trace paths apply unchanged.
- **Design note:** the deadline is read from the *signed token constraint* rather than a `Kernel.__init__` argument. This is the issue's explicitly-sanctioned "per-capability constraint" option, it is more secure (tamper-evident, bound to the grant), and it avoids growing `kernel/__init__.py`, which sits exactly at its shrink-only size-ratchet ceiling (541). A kernel-wide constructor default is a natural follow-up once #188 splits that module.

**#194 — Pooled HTTP client + response-size guard**
- `HTTPDriver` holds a single long-lived `httpx.AsyncClient` (connection pooling, configurable `httpx.Limits`) instead of building one per request, with an idempotent `aclose()` for shutdown. The `http_driver_demo` example now closes it in a `finally`.
- Optional `max_response_bytes` streams the body and aborts oversized responses with a `DriverError` before they are fully buffered (the firewall budget only applies *after* a `RawResult` exists).

**#197 — Defensive HTTP body parsing**
- A JSON endpoint returning a non-JSON body now raises a typed `DriverError` (content-type + redaction-safe snippet) instead of leaking `json.JSONDecodeError`.
- New `HTTPEndpoint.response_format` (`"json"` | `"text"`) supports text APIs deliberately.

## Why

The audit found live I-02 gaps on the execution path: a non-`DriverError` exception, a firewall failure, or a hung driver could each leave an invocation un-traced and/or leak a budget reservation. The HTTP driver opened a fresh connection pool per call and could buffer an unbounded body or crash the agent loop on a non-JSON response. This PR closes all four conservatively (every behaviour defaults to its prior semantics; the deadline and size cap are opt-in).

## How verified

Full repo gate (`make ci` → `fmt-check → lint → type → test → example`) in a clean venv — **exit 0**:
- `ruff format --check` — 119 files already formatted
- `ruff check` — All checks passed
- `mypy src/` (strict) — no issues in 62 source files
- `pytest` — **789 passed, 1 skipped**, branch coverage **93.91%** (floor 90)
- `make example` — all 13 examples complete, incl. `http_driver_demo.py` against a real local server

New/updated tests:
- `tests/test_driver_exec.py` (new) — fallback, deadline, and non-`DriverError` fault capture
- `tests/test_drivers.py` — rewritten HTTP tests for the pooled-client/streaming contract; new cases for non-JSON, `text` format, empty body, size-limit abort/allow, pooled reuse, `aclose`
- `tests/test_kernel.py` — non-`DriverError` audited + budget released; firewall-failure audited + budget released; deadline aborts a slow driver; invalid `invoke_timeout_s` rejected

## Tradeoffs / risks

- **#152 changes propagation semantics** for non-`DriverError` exceptions (now treated as failed attempts and re-wrapped as `DriverError` after the route plan is exhausted). This is intentional and central to the fix, but it's a behaviour change worth a maintainer eye.
- The HTTP driver now requires callers to own the client lifecycle (`aclose()`); the example demonstrates the pattern.

## Scope notes

Limited to the four issues above. **#170** (per-invocation rate-limit enforcement) was deliberately left out — it's a distinct rate-limiting/policy concern (`policy.py`/`rate_limit.py`) and would dilute this PR's focus; recommend it as a separate follow-up. A kernel-wide default deadline depends on the #188 module split.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018VdP3irZSbbPyyoVS1QQbD

---
_Generated by [Claude Code](https://claude.ai/code/session_018VdP3irZSbbPyyoVS1QQbD)_

### CODE REVIEW

**Comment by dgenio** [comment]
Addressed all five review findings in b8555bc:

1. **`max_response_bytes` bypassed on errors** (`http.py`) — the error path no longer does an unbounded `response.aread()`; it streams a bounded snippet (≤512 bytes, sliced to 200 chars) for the message, so the size guard holds on the failure path too.
2. **Failure trace recorded caller mode, not effective** (`_invoke.py`) — failure traces now record `effective_mode`, matching success and streaming traces.
3. **Failure trace hard-coded `driver_id=""`** (`_invoke.py`) — `record_failure_trace` takes an optional `driver_id`; the post-driver path passes the driver that ran, and `execute_with_fallback` now returns the last-attempted driver so the all-failed path attributes it too.
4. **CancelledError leaked budget / skipped audit** (`_invoke.py`) — the reservation is released in a single `finally` (which covers `CancelledError`, since it isn't an `Exception`) and a "invocation cancelled" failure trace is recorded before propagating. Added a test that cancels mid-invoke and asserts the reservation is freed and the cancel is audited.
5. **Stream trace dropped the real error reason** (`_stream.py`) — the trace now records the actual failure message (e.g. a timeout `DriverError`), redacted, instead of always `None` / "stream produced no chunks".

`make ci` is green (792 passed, 94.01% branch coverage).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/drivers/http.py [hunk 1]
```diff
@@ -79,47 +121,114 @@ async def execute(self, ctx: ExecutionContext) -> RawResult:
...
+                json=json_body,
+                headers=endpoint.headers,
+                timeout=effective_timeout,
+            ) as response:
+                if response.is_error:
+                    await response.aread()
+                    raise DriverError(
+                        f"HTTPDriver '{self._driver_id}': HTTP {response.status_code} "
+                        f"from {endpoint.url}: {response.text[:200]}"
                     )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/drivers/http.py:153]: Confirmed fixed on the current head (`057f410`). The error path no longer calls the unbounded `response.aread()` — it now uses `_read_error_snippet` (`drivers/http.py:174-194`), which streams via `aiter_bytes()` and stops at `max_bytes=512`, surfacing only a 200-char snippet. The `max_response_bytes` guard now holds on the non-2xx path too. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/kernel/_invoke.py [hunk 2]
```diff
@@ -343,10 +310,26 @@ async def perform_invoke(
...
             await kernel.budget.release(reserved_tokens)
+        err_msg = str(exc)
+        logger.warning("invoke_failure", extra={**log_ctx, "error": err_msg})
+        record_failure_trace(
+            action_id=action_id,
+            capability_id=token.capability_id,
+            principal_id=principal.principal_id,
+            token_id=token.token_id,
+            args=args,
+            response_mode=response_mode,
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/kernel/_invoke.py]: Confirmed fixed. Failure traces now record `response_mode=effective_mode` (`kernel/_invoke.py:280`), so the post-driver failure path is consistent with the success traces and the streaming path (which record the effective mode). Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/kernel/_invoke.py [hunk 3]
```diff
@@ -343,10 +310,26 @@ async def perform_invoke(
...
-        if not reservation_consumed and kernel.budget is not None and reserved_tokens is not None:
+    except Exception as exc:
+        if kernel.budget is not None and reserved_tokens is not None:
             await kernel.budget.release(reserved_tokens)
+        err_msg = str(exc)
+        logger.warning("invoke_failure", extra={**log_ctx, "error": err_msg})
+        record_failure_trace(
+            action_id=action_id,
+            capability_id=token.capability_id,
+            principal_id=principal.principal_id,
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/kernel/_invoke.py]: Confirmed fixed. `record_failure_trace` now accepts an optional `driver_id` (`kernel/_invoke.py:145`); the post-driver path passes the driver that ran, and `execute_with_fallback` returns the last-attempted driver so the all-failed path attributes it too (`_invoke.py:299`, `304`, `332`). Audit records now carry the implicated driver, matching the streaming path. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/kernel/_invoke.py [hunk 4]
```diff
@@ -293,7 +258,7 @@ async def perform_invoke(
     )
     downgraded = effective_mode != response_mode
     raw_result, used_driver_id, last_error, fell_back = await execute_with_fallback(
-        kernel._driver_map, plan, ctx=ctx, log_ctx=log_ctx
+        kernel._driver_map, plan, ctx=ctx, log_ctx=log_ctx, timeout=invoke_timeout
     )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/kernel/_invoke.py]: Confirmed fixed. `perform_invoke` now catches `asyncio.CancelledError` (`kernel/_invoke.py:331`) to record an "invocation cancelled" failure trace, and the reservation is released exactly once in a single `finally` (`_invoke.py:337-339`) that also covers cancellation (it isn't an `Exception`). A test cancels mid-invoke and asserts the reservation is freed and the cancel is audited, so the "audited on every exit" guarantee (I-02) now holds. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/kernel/_stream.py [hunk 5]
```diff
@@ -119,7 +123,16 @@ async def invoke_stream_impl(
...
+            try:
+                if invoke_timeout is None:
+                    raw = await fallback_driver.execute(ctx)
+                else:
+                    raw = await asyncio.wait_for(fallback_driver.execute(ctx), invoke_timeout)
+            except asyncio.TimeoutError as exc:
+                raise DriverError(
+                    f"Driver '{fallback_driver_id}' timed out after {invoke_timeout}s "
+                    f"for capability '{token.capability_id}'."
+                ) from exc
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/kernel/_stream.py:141]: Confirmed fixed. The stream now captures the real failure via `except Exception` (`kernel/_stream.py:164-168`) and the `finally` records the redacted error message instead of always `None`/"stream produced no chunks" (`_stream.py:170-175`). A timeout `DriverError`'s reason now lands in the `ActionTrace.error`. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_
